Microsoft.DotNet.Try.Jupyter/Rendering/RendererUtilities.cs

```diff
@@ -0,0 +1,185 @@
+{
+    internal static class RendererUtilities
+    {
+        public static bool IsStructured(Type sourceType)
+        {
+            var isStructured = !sourceType.IsPrimitive
+                               && ((sourceType.IsClass && sourceType != typeof(string))
+                                   || sourceType.IsValueType && !sourceType.IsEnum);
+            return isStructured;
+        }
+
+        public static string CreateTableHeaders(IEnumerable<MemberInfo> memberInfos, bool emptyFirstHeader)
+        {
+            var headersBuffer = new StringBuilder();
+            if (memberInfos?.Any() != false)
+            {
+                headersBuffer.AppendLine("\t<tr>");
+                if (emptyFirstHeader)
+                {
+                    headersBuffer.AppendLine("\t\t<th></th>");
+                }
+
+                foreach (var memberInfo in memberInfos)
+                {
+                    headersBuffer.AppendLine($"\t\t<th>{memberInfo.Name}</th>");
+                }
+                headersBuffer.AppendLine("\t</tr>");
+            }
+
+            return headersBuffer.ToString();
+        }
+
+        public static IEnumerable<MemberInfo> GetAccessors(Type sourceType)
+        {
+            if (IsStructured(sourceType))
+            {
+                var props = sourceType.GetProperties(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static).OfType<MemberInfo>();
+                var fields = sourceType.GetFields(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static).OfType<MemberInfo>();
+
+                return props.Concat(fields);
+            }
+
+            return Enumerable.Empty<MemberInfo>();
+        }
+
+        public static string CreateTableRowsFromValues(IEnumerable<MemberInfo> memberInfos, IEnumerable<KeyValuePair<object,object>> source, IRenderingEngine engine,
+           bool emptyFirstCell = false)
+        {
+            var nullRendering = new PlainTextRendering("null");
+            var rowsBuffer = new StringBuilder();
+
+
+            foreach (var (key,element) in source.Select(e => (e.Key,e.Value)))
+            {
+
+                rowsBuffer.AppendLine("\t<tr>");
+                if (emptyFirstCell)
+                {
+                    rowsBuffer.AppendLine("\t\t<td></td>");
+                }
+
+                var keyRenderer = engine.TryFindRenderer(key.GetType());
+                var keyRendering = keyRenderer.Render(key, engine);
+                rowsBuffer.AppendLine($"\t\t<td>{keyRendering?.Content ?? string.Empty}</td>");
+                CreateTableRow(memberInfos, engine, element, nullRendering, rowsBuffer);
+                rowsBuffer.AppendLine("\t</tr>");
+            }
+
+            return rowsBuffer.ToString();
+        }
+
+        private static void CreateTableRow(IEnumerable<MemberInfo> memberInfos, IRenderingEngine engine, object element,
+            PlainTextRendering nullRendering, StringBuilder rowsBuffer)
```

Comment:
Can I pass a PlainTextRendering here which is not null ? Both the functions that call it pass a null here, so we can create the nullRendering inside here only ?

Reply:
null rendering is the rendering to use if null. so the caller can decide